File: Lab1/lessons.py
from db import *
from app import app
from sqlalchemy import and_
from sqlalchemy.orm import Session 
from fastapi import Depends, Body, status 
from fastapi.responses import JSONResponse
from datetime import date

# ---------------------------------------------------
@app.get("/api/lessons") 
# За допомогою класу Depends() у функцію передається результат функції get_db()
def get_lessons(db: Session = Depends(get_db)): 
    """
    перелік занять
    """
    # отримуємо  список  об'єктів Lesson
    return db.query(Lesson).all() 

# ---------------------------------------------------
@app.get("/api/lessons/{id}") 
def get_lesson(id: int, db: Session = Depends(get_db)): 
    """
    заняття за заданим id
    """
    # отримуємо заняття за id 
    lesson = db.query(Lesson).filter(Lesson.id == id).first() 
    # якщо не знайдено, відправляємо статусний код і повідомлення про помилку 
    if lesson == None:   
        return JSONResponse(status_code = status.HTTP_404_NOT_FOUND, content={ "message": "Заняття не знайдено"}) 
    #якщо заняття знайдено, відправляємо його 
    return lesson

# ---------------------------------------------------    
@app.post("/api/lessons")
def create_lesson(idSubject: int = Body(embed=True),
                  idTeacher: int = Body(embed=True),
                  dateLesson: date = Body(embed=True),
                  idJournal: int = Body(embed=True),
                  theme: str = Body(embed=True),
                  maxGrade: int = Body(default = None, embed=True),
                  db: Session = Depends(get_db)):
    """
    створення нового заняття
    """
    lesson = Lesson(
        idSubject = idSubject, 
        idTeacher = idTeacher,
        dateLesson = dateLesson,
        idJournal = idJournal,
        theme = theme,
        maxGrade = maxGrade)
    db.add(lesson) 
    db.commit() 
    db.refresh(lesson) 
    return lesson 

# ---------------------------------------------------   
@app.post("/api/lessons/find") 
def find_lesson(data  = Body(), db: Session = Depends(get_db)):
    """
    пошук заняття за заданими параметрами
    """
    lesson = db.query(Lesson).filter(and_(Lesson.idJournal == data["idJournal"], Lesson.idSubject == data["idSubject"], Lesson.idTeacher == data["idTeacher"], Lesson.dateLesson == data["dateLesson"])).first()
    # якщо не заняття за обраними критеріями не знайдено, відправляємо статусний код і повідомлення про помилку 
    if lesson == None:   
        return JSONResponse(status_code = status.HTTP_404_NOT_FOUND, content={ "message": "Заняття не знайдено"}) 
    #якщо заняття знайдено, відправляємо його 
    return lesson
   
# ---------------------------------------------------   
@app.put("/api/lessons") 
def edit_lesson(id: int = Body(embed=True),
                idSubject: int = Body(embed=True),
                idTeacher: int = Body(embed=True),
                dateLesson: date = Body(embed=True),
                idJournal: int = Body(embed=True),
                theme: str = Body(embed=True),
                maxGrade: int = Body(default = None, embed=True),
                db: Session = Depends(get_db)):
    """
    зміна даних про заняття
    """
    # отримуємо заняття за id 
    lesson = db.query(Lesson).filter(Lesson.id == id).first() 
    # якщо не знайдено, відправляємо статусний код і повідомлення про помилку 
    if lesson == None:  
        return JSONResponse(status_code = status.HTTP_404_NOT_FOUND, content={ "message": "Заняття не знайдено"}) 
    # якщо заняття знайдено, змінюємо його дані і відправляємо назад клієнту 
    lesson.idSubject = idSubject
    lesson.idTeacher = idTeacher
    lesson.dateLesson = dateLesson
    lesson.idJournal = idJournal
    lesson.theme = theme
    lesson.maxGrade = maxGrade

    db.commit() # зберігаємо зміни  
    db.refresh(lesson) 
    return lesson 
   
# ---------------------------------------------------   
@app.delete("/api/lessons/{id}") 
def delete_lesson(id, db: Session = Depends(get_db)):
    """
    видалення заняття
    """
    # отримуємо заняття за id 
    lesson = db.query(Lesson).filter(Lesson.id == id).first() 
    
    # якщо не знайдено, відправляємо статусний код і повідомлення про помилку 
    if lesson == None: 
        return JSONResponse( status_code = status.HTTP_404_NOT_FOUND, content={ "message": "Заняття не знайдено"}) 
    
    # якщо заняття знайдено, видаляємо його 
    db.delete(lesson)  # видаляємо об'єкт 
    db.commit()     # зберігаємо зміни 
    return lesson 

File: Lab5/wp1/journal/static/js/lessons.js
/* global bootstrap: false */
// ---------------------------------------------------
// Пошук заняття за заданими параметрами
const findLesson = async () => {

  try {
    // надсилаємо запит 
    const response = await fetch(`/api/lessons/find`, {
      method: "POST",
      headers: { 
        "Accept": "application/json", 
        "Content-Type": "application/json",
        'X-CSRFToken': getCookie('csrftoken') },  
      // body - об'єкт класу Lesson     
      body: JSON.stringify({  
        "idJournal": formLesson.journal.value, 
        "idTeacher": getCookie("user_id"),
        "idSubject":  formLesson.subject.value, 
        "dateLesson": formLesson.date_lesson.value
      }),
      credentials: 'include'  // важливо, щоб куки передавались!       
    });

    // заняття знайдено
    if (response.ok) {
      const lesson = await response.json();
      console.log("lesson=", lesson);
      // У компоненти форм записуємо значення id та теми заняття
      formLesson.idLesson.value = lesson.id;  
      formRating.theme.value = lesson.theme;
      formRating.max_grade.value = lesson.maxGrade;

      // Отримуємо та відображаємо рейтинг студентів за заданим lesson.id
      await getRating(lesson.id);        
      // Помилки 
    } else {
      // Аналізуємо статус відповіді
      if (response.status === 404) {
        showToast("Увага!", "Заняття не знайдено.", 1);
        // відображаємо кнопку "Нове заняття"
        formLesson.btnNewLesson.hidden = false;               
      } else {   
        showToast("Помилка!", `Помилка запиту даних про заняття, статус: ${response.status}`, 2);        
        console.log("findLesson:", response);
      }
    }

  }
  catch (error) {
    console.log("findLesson:", error);
    showToast("Помилка!", error, 2);         
  }
}

// ---------------------------------------------------
// Додавання нового заняття
const addLesson = async () => {

  try {
    // надсилаємо запит 
    const response = await fetch(`/api/lessons/new`, {
      method: "POST",
      headers: { 
        "Accept": "application/json", 
        "Content-Type": "application/json",
        'X-CSRFToken': getCookie('csrftoken') },
      // body - об'єкт класу Lesson     
      body: JSON.stringify({  
        "idJournal": formLesson.journal.value, 
        "idTeacher": getCookie("user_id"),
        "idSubject":  formLesson.subject.value, 
        "dateLesson": formLesson.date_lesson.value,
        "theme": formRating.theme.value,
        "maxGrade": formRating.max_grade.value        
      })       
    });

    if (response.ok) {
      const lesson = await response.json();
      formLesson.idLesson.value = lesson.id;
      console.log("added lesson=", lesson);    

      // Додаємо рейтинг студентів за заданим lesson.id
      await addRating(lesson.id);

      // Помилки 
    } else {
      showToast("Помилка!", `Помилка створення нового заняття, статус: ${response.status}`, 2);         
      console.log("addLesson:", response);
    }

  }
  catch (error) {
    console.log("addLesson:", error);
    showToast("Помилка!", `Помилка запиту створення нового заняття: ${error}`, 2);        
  }
}

// ---------------------------------------------------
// Коригування заняття
const editLesson = async () => {

  try {
    // надсилаємо запит 
    const response = await fetch(`/api/lessons/edit`, {
      method: "PUT",
      headers: { 
        "Accept": "application/json", 
        "Content-Type": "application/json",
        'X-CSRFToken': getCookie('csrftoken') },
      // body - об'єкт класу Lesson     
      body: JSON.stringify({  
        "id": formLesson.idLesson.value,  
        "idJournal": formLesson.journal.value, 
        "idTeacher": getCookie("user_id"),
        "idSubject":  formLesson.subject.value, 
        "dateLesson": formLesson.date_lesson.value,
        "theme": formRating.theme.value,        
        "maxGrade": formRating.max_grade.value
      })       
    });

    if (response.ok) {
      const lesson = await response.json();
      console.log("edited lesson=", lesson);    

      // Коригуємо рейтинг студентів за заданим lesson.id
      await editRating(lesson.id);

      // Помилки 
    } else {
      showToast("Помилка!", `Помилка коригування заняття, статус: ${response.status}`, 2);             
      console.log("editLesson:", response);
    }

  }
  catch (error) {
    console.log("editLesson:", error);
    showToast("Помилка!", `Помилка запиту на коригування заняття: ${error}`, 2);        
  }
}

// ---------------------------------------------------
// Видалення заняття
const deleteLesson = async () => {
  // Спочатку видаляємо рейтинг студентів за заданим formLesson.idLesson.value
  const resultDelRating = await deleteRating(formLesson.idLesson.value);
  if (!resultDelRating)
    return;

  try {
    // надсилаємо запит на видалення заняття
    const response = await fetch(`/api/lessons/delete/${formLesson.idLesson.value}`, {
      headers: { 
        "Accept": "application/json", 
        "Content-Type": "application/json",
        'X-CSRFToken': getCookie('csrftoken') },      
      method: "DELETE"   
    });

    if (response.ok) {
      console.log("Заняття з id=", formLesson.idLesson.value, "видалено");    
      showToast("Повідомлення", `Заняття успішно видалено. <br>Зміни записані до БД.`, 0);        
      formLesson.subject.disabled = false;
      formLesson.subject.selectedIndex = -1;
      formLesson.journal.disabled = false;
      formLesson.journal.selectedIndex = -1;      
      formLesson.date_lesson.disabled = false;
      formLesson.btnFindLesson.hidden = false;
      formLesson.btnFindLesson.disabled = true;
      formLesson.btnNewLesson.hidden = true;
      formLesson.idLesson.value = "";

      document.getElementById("idDivRating").hidden = true;
      // Помилки 
    } else {
      showToast("Помилка!", `Помилка видалення заняття, статус: ${response.status}`, 2);        
      console.log("deleteLesson: ", response);
    }

  }
  catch (error) {
    console.log("deleteLesson:", error);
    showToast("Помилка!", `Помилка запиту на видалення заняття: ${error}`, 2);  
  }
}
